### ISSUE TITLE
Make JSON a union type

### ISSUE THREAD

**Issue by SupunS** [ballerina-platform/ballerina-lang#4295]
json type should be one of: number, string, boolean, null, array, map<json>

**Comment by SupunS** [comment]
Related to: https://github.com/ballerinalang/ballerina/issues/4142

**Comment by SupunS** [comment]
@sanjiva, @sameera-jayasoma, 
As per the discussion we had with regards to this, we came to a conclusion to change the underlying implementation to ballerina types, rather than a generic JSONNode implementation. i.e: a BInteger to represent an int type JSON, a BMap to represent object type JSON, etc. So that even during a type cast (say JSON to map), it will be just a change the pointer, and wont do any byte level conversion.

But after further analyzing, I realized this is not feasible. Reason is,  suppose I create an int array like `int[] a = [1, 2, 3]`. In the BVM this is stored as java int[]. On the other hand, if we make JSON a union type and if I initialize a JSON using the same array: `json j = [1, 2, 3]`. this will get stored as BInteger[]. So the two problems here is:
- ballerina int[] are stored without boxing, but JSON type arrays are store with boxing.
- Suppose I cast the int[] to JSON like: `json j = (json) a;` This is perfectly valid since JSON can hold int arrays. But now I can add non-int type values to `j`, since its a JSON. That means we cannot maintain the reference between `j` and `a` --> which implies this cannot be a cast operation.

The same issue exists with representing JSON object type nodes with BMap. i.e: map tp json can never be a cast, its always needs to be a conversion. 

So AFAIU, we can't really make them a union type, even if we do so, we dont get any gain in terms of functionality or in performance (we may lose some performance since we have to do boxing and unboxing for value types). However, I have made "typeof j" work correctly, as per #4142, without chaning the underlying representation.

Appreciate your thoughts.
